Remove usages of MailPoet\Tasks\Sending — Part 1

JanJakes opened the pull request:
## Description

This PR removes the first batch of usages of `MailPoet\Tasks\Sending`.

## Code review notes

In the first part, this PR removes some fairly simple usages of `MailPoet\Tasks\Sending`. The second part removes the usages from `MailPoet\Cron\Workers\Scheduler` and does so gradually, method by method — the commits tell the story.

## QA notes

Please, verify that all types of email sending in the plugin work and that the emails render correcdtly:

1. Newsletters (immediate as well as scheduling).
2. Post notifications.
3. Welcome emails.
4. Transactional emails.
5. Automation emails (including transactional).
6. Re-engagement emails.
7. Stat notifications.

We have a fairly good test coverage for email sending, but it is worth verifying.

## Linked PRs

_N/A_

## Linked tickets

[MAILPOET-4375]

## After-merge notes

_N/A_

## Tasks

- [x] I followed [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers) for translations
- [x] I added sufficient test coverage
- [x] I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes


[MAILPOET-4375]: https://mailpoet.atlassian.net/browse/MAILPOET-4375?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

JanJakes committed:
Remove usages of MailPoet\Tasks\Sending from Help page

[MAILPOET-4375]

JanJakes committed:
Remove usages of MailPoet\Tasks\Sending from reengagement scheduler

[MAILPOET-4375]

JanJakes committed:
Remove usages of MailPoet\Tasks\Sending from automation email scheduler

[MAILPOET-4375]

JanJakes committed:
Remove usages of MailPoet\Tasks\Sending from stats notification worker

[MAILPOET-4375]

JanJakes committed:
Remove usages of MailPoet\Tasks\Sending from welcome scheduler

[MAILPOET-4375]

JanJakes committed:
Remove usages of MailPoet\Tasks\Sending from renderer

[MAILPOET-4375]

JanJakes committed:
Enable usage of both Doctrine and legacy task/queue entities in scheduler loop

[MAILPOET-4375]

JanJakes committed:
Implement scheduler task, queue, and subscribers removal with Doctrine

[MAILPOET-4375]

JanJakes committed:
Rewrite Scheduler::processWelcomeNewsletter() using Doctrine

[MAILPOET-4375]

JanJakes committed:
Rewrite Scheduler::processPostNotificationNewsletter() using Doctrine

[MAILPOET-4375]

JanJakes committed:
Rewrite Scheduler::processScheduledStandardNewsletter() using Doctrine

[MAILPOET-4375]

JanJakes committed:
Rewrite Scheduler::processScheduledAutomaticEmail() using Doctrine

[MAILPOET-4375]

JanJakes committed:
Rewrite Scheduler::processReEngagementEmail() using Doctrine

[MAILPOET-4375]

JanJakes committed:
Rewrite Scheduler::processScheduledAutomationEmail() using Doctrine

[MAILPOET-4375]

JanJakes committed:
Rewrite Scheduler::processScheduledTransactionalEmail() using Doctrine

[MAILPOET-4375]

JanJakes committed:
Remove no longer needed Scheduler code

[MAILPOET-4375]

JanJakes committed:
Remove simple usages of MailPoet\Tasks\Sending from integration tests

[MAILPOET-4375]

JanJakes committed:
Rewrite ShedulerTest using Doctrine

[MAILPOET-4375]

JanJakes committed:
Refresh task entity when any subscribers were added from segments

This is to keep $task->getSubscribers() up-to-date.

[MAILPOET-4375]

JanJakes committed:
Use orphan removal for scheduled task subscriber entity

It is a connecting entity between task and subscriber, and it makes no sense
to keep without either of them. This could prevent some partial-delete bugs.

[MAILPOET-4375]

JanJakes commented on mailpoet/lib/Newsletter/Scheduler/WelcomeScheduler.php:
```diff
@@ -118,28 +119,31 @@ public function createWelcomeNotificationSendingTask(NewsletterEntity $newslette
+    $queue = new SendingQueueEntity();
+    $queue->setTask($task);
+    $queue->setNewsletter($newsletter);
+    $queue->setSubscribers((string)$subscriberId);
```
Do you happen to know what's this column for? Seems to be a string (comma separated IDs?), but couldn't find any usages of `getSubscribers()` on the queue entity. It's the task entity where we keep the relationships, but I'm not sure what the queue subscribers column is for, and whether I should set it here or not.

rodrigoprimo replied:
I don't know why we have a subscribers column in wp_mailpoet_sending_queues. Maybe it is a field that is not used anymore?

> whether I should set it here or not

Do you have any indication that we should set it? I'm inclined to think we don't need to set it. On a quick check, it seems that the code, even before the refactor that I did in the subscriber related methods of the Sending class, does not set the `subscribers` column in `createWelcomeNotificationSendingTask()`.

JanJakes replied:
You're right, I think I added that by mistake instead of setting scheduled task subscribers to the `$task`, then noticed those were missing, and then added those too. This indeed seems to do nothing.

I checked all usages and removed them: https://github.com/mailpoet/mailpoet/pull/5241/commits/388b078725a12f93aa95aa0bb736862926640bf7

rodrigoprimo commented on mailpoet/lib/Cron/Workers/StatsNotifications/Worker.php:
```diff
@@ -96,7 +96,7 @@ public function process($timer = false) {
     // Cleanup potential orphaned task created due bug MAILPOET-3015
     $this->repository->deleteOrphanedScheduledTasks();
-    foreach ($this->repository->findScheduled(Sending::RESULT_BATCH_SIZE) as $statsNotificationEntity) {
+    foreach ($this->repository->findScheduled(self::BATCH_SIZE) as $statsNotificationEntity) {
```
Can we remove `Sending::RESULT_BATCH_SIZE` since it doesn't seem to be used anymore?

JanJakes replied:
Removed: https://github.com/mailpoet/mailpoet/pull/5241/commits/45efeda4b7e97d829e8453bc74a250597edd9078

rodrigoprimo commented on mailpoet/tests/integration/Cron/Workers/SchedulerTest.php:
```diff
@@ -582,15 +584,19 @@ public function testItCanProcessPostNotificationNewsletter() {
+      $this->entityManager->getClassMetadata(NewsletterEntity::class)->getTableName(),
+      ['id' => $newsletter->getId()]
+    );
+    $this->entityManager->clear();
```
It is not necessarily a problem, but it is not obvious to me why we need to call `clear()` here. The test passes without it.

JanJakes replied:
Oh, good catch! It was a bug, this test (and the two others) passed even without the delete! It was because there was neither a scheduled task subscriber, not a matching segment, so it deleted the queue in any way. I fixed all of them, and now the deletes and `->clear()` calls are needed.

Fix: https://github.com/mailpoet/mailpoet/pull/5241/commits/fcbd3c81023885952dc189966779335441702a89

rodrigoprimo commented on mailpoet/tests/integration/Cron/Workers/SchedulerTest.php:
```diff
@@ -768,149 +782,112 @@ public function testItProcessesScheduledJobsWhenNewsletterIsScheduled() {
+      $this->entityManager->getClassMetadata(SubscriberEntity::class)->getTableName(),
+      ['id' => $subscriber->getId()]
+    );
+    $this->entityManager->clear();
```
Same here, it doesn't seem like we need to call `clear()` here.

JanJakes replied:
Fix: https://github.com/mailpoet/mailpoet/pull/5241/commits/810e28f645b5b45eaf25b96ec3999b60ae98f0dd

rodrigoprimo commented on mailpoet/tests/integration/Cron/Workers/SchedulerTest.php:
```diff
@@ -768,149 +782,112 @@ public function testItProcessesScheduledJobsWhenNewsletterIsScheduled() {
+      $this->entityManager->getClassMetadata(SubscriberEntity::class)->getTableName(),
+      ['id' => $subscriber->getId()]
+    );
+    $this->entityManager->clear();
```
One more

JanJakes replied:
Fix: https://github.com/mailpoet/mailpoet/pull/5241/commits/6b2120a9ce8bedd9eb894f282b8be34ae4ba7391

rodrigoprimo commented on mailpoet/lib/Newsletter/Sending/SendingQueuesRepository.php:
```diff
@@ -205,4 +205,20 @@ public function saveFilterSegmentMeta(SendingQueueEntity $queue, SegmentEntity $
     $this->flush();
   }
+
+  public function updateCounts(SendingQueueEntity $queue, ?int $count = null): void {
```
Is it on purpose that this method doesn't call `flush()`? I'm not sure if there is a convention for that or not, but I would expect it to persist the changes to the database. That is what the other methods in this repository class seem to be doing.

JanJakes replied:
From the current usages it didn't seem necessary, but on the other hand, I agree it's safer to flush here.

Added: https://github.com/mailpoet/mailpoet/pull/5241/commits/08d95841d5596566e095a24a66306c0e5d453212

rodrigoprimo reviewed (commented):
Great work! The changes look good to me. I left a few minor comments for us to discuss.

JanJakes committed:
Fix and improve SchedulerTest::testItDeletesQueueDuringProcessingWhenNewsletterNotFound()

Having no segment and no task subscriber caused the test pass even without removing the newsletter.

[MAILPOET-4375]

JanJakes committed:
Fix and improve SchedulerTest::testItDeletesScheduledAutomaticEmailWhenUserDoesNotExist()

Having no segment and no task subscriber caused the test pass even without removing the subscriber.

[MAILPOET-4375]

JanJakes committed:
Fix and improve SchedulerTest::testItDeletesScheduledAutomationEmailWhenUserDoesNotExist()

Having no segment and no task subscriber caused the test pass even without removing the subscriber.

[MAILPOET-4375]

JanJakes committed:
Simplify and improve SchedulerTest::testItCanDeleteQueueWhenDeliveryIsSetToImmediately()

The entity fetching for Doctrine is no longer needed, added more checks.

[MAILPOET-4375]

JanJakes committed:
Simplify and improve SchedulerTest::testItCanRescheduleQueueDeliveryTime()

The entity fetching for Doctrine is no longer needed, added more checks.

[MAILPOET-4375]

JanJakes committed:
Fix and improve SchedulerTest::testItFailsWPSubscriberVerificationWhenSubscriberIsNotAWPUser()

This test would pass even with correct WP user ID, as there was a role mismatch ("author" vs. "editor").

[MAILPOET-4375]

JanJakes committed:
Improve SchedulerTest::testItDoesNotProcessWelcomeNewsletterWhenThereAreNoSubscribersToProcess()

[MAILPOET-4375]

JanJakes committed:
Improve SchedulerTest::testItProcessesWelcomeNewsletterWhenSubscriberIsVerified()

Added NewsletterEntity::TYPE_WELCOME to make it correspond better to the fucntion name,
plus fixed a comment.

[MAILPOET-4375]

JanJakes committed:
Improve SchedulerTest::testItProcessesWelcomeNewsletterWhenWPUserIsVerified()

Added NewsletterEntity::TYPE_WELCOME to make it correspond better to the fucntion name,
plus fixed a comment.

[MAILPOET-4375]

JanJakes committed:
Improve SchedulerTest::testItFailsMailpoetSubscriberVerificationWhenSubscriberDoesNotExist()

[MAILPOET-4375]

JanJakes committed:
Improve SchedulerTest::testItFailsMailpoetSubscriberVerificationWhenSubscriberIsNotInSegment()

[MAILPOET-4375]

JanJakes committed:
Improve SchedulerTest::testItProcessesScheduledStandardNewsletter()

Added NewsletterEntity::TYPE_STANDARD to make it correspond better to the fucntion name.

[MAILPOET-4375]

JanJakes committed:
Simplify SchedulerTest::testItFailsToProcessPostNotificationNewsletterWhenSubscribersNotInSegment()

There is no need to mock "deleteQueueOrUpdateNextRunDate" to make the subscriber-segment check work.

[MAILPOET-4375]

JanJakes committed:
Simplify SchedulerTest::testItDeletesQueueDuringProcessingWhenNewsletterIsSoftDeleted()

[MAILPOET-4375]

JanJakes committed:
Extract creating scheduler mock to a method

[MAILPOET-4375]

JanJakes committed:
Remove unnecessary code from SchedulerTest

[MAILPOET-4375]

JanJakes committed:
Remove MailPoet\Tasks\Sending::BATCH_SIZE

[MAILPOET-4375]

JanJakes committed:
Call flush() in SendingQueuesRepository::updateCounts()

[MAILPOET-4375]

JanJakes committed:
Remove SendingQueueEntity::$subscribers

This doesn't seem to be used anywhere. For scheduled task subscribers, we have
ScheduledTaskEntity::$subscribers collection.

[MAILPOET-4375]

JanJakes commented:
@rodrigoprimo Thanks for the review and your observations! It made me double-check all of the scheduler tests one by one and reveal more issues and unnecessary code. I fixed, improved, and simplified more of the tests, starting from https://github.com/mailpoet/mailpoet/pull/5241/commits/fcbd3c81023885952dc189966779335441702a89.

Aschepikov reviewed (approved):
GJ!
